README.md
# Rack::Multiplexer
Provides a simple router & dispatcher for Rack applications as a Rack application.  
The routing algorithm has only O(1) time complexity in Ruby level because all routes are compiled into one Regexp.

## Installation
```
gem install rack-multiplexer
```

## Usage
For `rackup`.

```ruby
# config.ru
require "rack-multiplexer"

multiplexer = Rack::Multiplexer.new
multiplexer.get("/a", ->(env) { [200, {}, ["a"]] })
multiplexer.get("/b", ->(env) { [200, {}, ["b"]] })
multiplexer.put("/c", ->(env) { [200, {}, ["c"]] })
multiplexer.get("/d/:e", ->(env) { [200, {}, [env["rack.request.query_hash"]["e"]]] })

run multiplexer
```

## DSL
The block is with you, always.

```ruby
# config.ru
require "rack-multiplexer"

run Rack::Multiplexer.new {
  get "/a" do
    [200, {}, ["a"]]
  end

  get "/b/:c" do
    [200, {}, ["d"]]
  end
}
```


CHANGELOG.md
## 0.0.8
* Support PATCH request method

## 0.0.7
* Remove delendency on rspec

## 0.0.6
* Allow to be a rack application too

## 0.0.5
* Fix root path delegation

## 0.0.4
* Fix regexp meta-characters in path pattern

## 0.0.3
* Introduce regexp compilation algorithm

## 0.0.2
* Support block-style DSL

## 0.0.1
* 1st release


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "rack/multiplexer"

RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true
  config.filter_run :focus
end


spec/rack/multiplexer_spec.rb
require "spec_helper"
require "active_support/core_ext/object/to_query"
require "active_support/core_ext/object/try"

describe Rack::Multiplexer do
  let(:application) do
    ->(env) {
      [
        200,
        {},
        ["#{env['PATH_INFO']}?#{env['rack.request.query_hash'].try(:to_param)}&#{env['QUERY_STRING']}"]
      ]
    }
  end

  let(:env) do
    {
      "SCRIPT_NAME" => "",
      "SERVER_NAME" => "localhost",
      "SERVER_PORT" => "80",
      "rack.input" => StringIO.new(""),
    }
  end

  describe "#call" do
    context "with unrelated path request" do
      it "sends request to default not found application" do
        multiplexer = described_class.new
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/a")).should == [
          404,
          { "Content-Type" => "text/plain", "Content-Length" => "0" },
          [""],
        ]
      end
    end

    context "with custom not found application" do
      it "sends request to custom not found application" do
        multiplexer = described_class.new(->(env) { [404, {}, ["custom"]] })
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/a")).should == [
          404,
          {},
          ["custom"],
        ]
      end
    end

    context "with unrelated method" do
      it "delegates to not found application" do
        multiplexer = described_class.new
        multiplexer.post("/a", application)
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/a"))[0].should == 404
      end
    end

    context "with related method" do
      it "delegates to registered application" do
        multiplexer = described_class.new
        multiplexer.get("/a", application)
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/a"))[0].should == 200
      end
    end

    context "with 2 registered routes" do
      it "delegates to first-registered application" do
        multiplexer = described_class.new
        multiplexer.get("/a", application)
        multiplexer.get("/:any", application)
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/a"))[2][0].should == "/a?&"
      end
    end

    context "with path parameters pattern" do
      it "delegates with rack.request.query_hash" do
        multiplexer = described_class.new
        multiplexer.get("/:any", application)
        multiplexer.get("/a", application)
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/a"))[2][0].should == "/a?any=a&"
      end
    end

    context "with duplicated params in path & query string" do
      it "delegates with rack.request.query_hash & QUERY_STRING" do
        multiplexer = described_class.new
        multiplexer.get("/:any", application)
        multiplexer.call(
          env.merge(
            "REQUEST_METHOD" => "GET",
            "PATH_INFO" => "/a",
            "QUERY_STRING" => "any=b"
          )
        )[2][0].should == "/a?any=a&any=b"
      end
    end

    context "with block routing" do
      it "delegates to given block as rack application" do
        multiplexer = described_class.new
        multiplexer.get("/a") {|env| [200, {}, ["a"]] }
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/a"))[0].should == 200
      end
    end

    context "with any routing" do
      it "matches any method" do
        multiplexer = described_class.new
        multiplexer.any("/a", application)
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/a"))[0].should == 200
        multiplexer.call(env.merge("REQUEST_METHOD" => "POST", "PATH_INFO" => "/a"))[0].should == 200
        multiplexer.call(env.merge("REQUEST_METHOD" => "HEAD", "PATH_INFO" => "/a"))[0].should == 200
      end
    end

    context "with path including '.'" do
      it "matches exactly" do
        multiplexer = described_class.new
        multiplexer.get("/a.*") {|env| [200, {}, ["ok"]] }
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/a.*"))[0].should == 200
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/aaa"))[0].should == 404
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/abc"))[0].should == 404
      end
    end

    context "with root path" do
      it "delegates to registered application" do
        multiplexer = described_class.new
        multiplexer.get("/", application)
        multiplexer.call(env.merge("REQUEST_METHOD" => "GET", "PATH_INFO" => "/"))[0].should == 200
      end
    end
  end
end


lib/rack/multiplexer.rb
require "rack/multiplexer/version"
require "rack/request"

# Provides a simple router & dispatcher for Rack applications as a Rack application.
# The routing algorithm has only O(1) time complexity because all routes are compiled into one Regexp.
#
# Example:
#
#   # config.ru
#   multiplexer = Rack::Multiplexer.new
#   multiplexer.get("/a", ->(env) { [200, {}, ["a"]] })
#   multiplexer.get("/b", ->(env) { [200, {}, ["b"]] })
#   multiplexer.put("/c", ->(env) { [200, {}, ["c"]] })
#   multiplexer.get("/d/:e", ->(env) { [200, {}, [env["rack.request.query_hash"]["e"]]] })
#   run multiplexer
#
module Rack
  class Multiplexer
    def initialize(not_found_application = nil, &block)
      @not_found_application = not_found_application
      instance_eval(&block) if block
    end

    def call(env)
      path = env["PATH_INFO"]
      (
        routes[env["REQUEST_METHOD"]].find(path) ||
        routes["ANY"].find(path) ||
        not_found_application
      ).call(env)
    end

    def get(pattern, application = nil, &block)
      append("GET", pattern, application || block)
    end

    def post(pattern, application = nil, &block)
      append("POST", pattern, application || block)
    end

    def put(pattern, application = nil, &block)
      append("PUT", pattern, application || block)
    end

    def patch(pattern, application = nil, &block)
      append("PATCH", pattern, application || block)
    end

    def delete(pattern, application = nil, &block)
      append("DELETE", pattern, application || block)
    end

    def any(pattern, application, &block)
      append("ANY", pattern, application || block)
    end

    def append(method, pattern, application)
      routes[method] << Route.new(pattern, application)
    end

    def routes
      @routes ||= Hash.new {|hash, key| hash[key] = Routes.new }
    end

    def not_found_application
      @not_found_application ||= ->(env) {
        [
          404,
          {
            "Content-Type" => "text/plain",
            "Content-Length" => "0",
          },
          [""],
        ]
      }
    end

    class Routes
      def initialize
        @routes = []
      end

      def find(path)
        if regexp === path
          @routes.size.times do |i|
            return @routes[i] if Regexp.last_match("_#{i}")
          end
        end
      end

      def <<(route)
        @routes << route
      end

      private

      def regexp
        @regexp ||= begin
          regexps = @routes.map.with_index {|route, index| /(?<_#{index}>#{route.regexp})/ }
          /\A#{Regexp.union(regexps)}\z/
        end
      end
    end

    class Route
      PLACEHOLDER_REGEXP = /:(\w+)/

      attr_reader :regexp

      def initialize(pattern, application)
        @application = application
        @regexp, @keys = compile(pattern)
      end

      def call(env)
        request = Rack::Request.new(env)
        data = @regexp.match(env["PATH_INFO"])
        (data.size - 1).times {|i| request.update_param(@keys[i], data[i + 1]) }
        @application.call(request.env)
      end

      def match?(path)
        @regexp === path
      end

      def compile(pattern)
        keys = []
        segments = []
        pattern.split("/").each do |segment|
          segments << Regexp.escape(segment).gsub(PLACEHOLDER_REGEXP, "([^#?/]+)")
          if key = Regexp.last_match(1)
            keys << key
          end
        end
        return Regexp.new(segments.any? ? segments.join(?/) : ?/), keys
      end
    end
  end
end
